### ISSUE TITLE
Testing

### ISSUE THREAD

**Issue by videlec** [rpy2/rpy2#670]
- Having `pytest` as an install dependency is very inconvenient for packaging. Would you consider removing it from the dependency list? (would be solved by #1)
- Is there any deep reason why tests are included inside the `rpy2` module and not in a separated `test/` or `tests/` folder? Most users do not want to install them.

**Comment by lgautier** [comment]
> Having pytest as an install dependency is very inconvenient for packaging. Would you consider removing it from the dependency list? (would be solved by #1)

Would you be able to elaborate more on `pytest`'s specific inconvenience compared to  'cffi', 'pytz', etc..., all of which are rpy2 dependencies  ? Also, PR #1 adds `pytest-runner` as a dependency, so if you are after shedding dependencies this might not help.

Making `pytest` an optional dependency is not necessarily off-the-table, but I'd to understand more how and why it matters.

> Is there any deep reason why tests are included inside the rpy2 module and not in a separated test/ or tests/ folder? Most users do not want to install them.

Most users ? I am happy to look at data about rpy2 users if you are able to share them.
There is of course this (https://github.com/rpy2/rpy2/issues/542), but otherwise I do think that it is a good idea to ship the tests because of the dependency on R. The R installation, or the R in the PATH, can change yet a user may want to check whether it is still clearing the tests.

**Comment by videlec** [comment]
> > Having pytest as an install dependency is very inconvenient for packaging. Would you consider removing it from the dependency list? (would be solved by #1)
> 
> Would you be able to elaborate more on `pytest`'s specific inconvenience compared to 'cffi', 'pytz', etc..., all of which are rpy2 dependencies ? Also, PR #1 adds `pytest-runner` as a dependency, so if you are after shedding dependencies this might not help.
>
> Making `pytest` an optional dependency is not necessarily off-the-table, but I'd to understand more how and why it matters.

Simply, `pytest` is not a runtime dependency of `rpy2` whereas `cffi` and `pytz` are. With the current status, it forces me (as a SageMath package manager) to also package `pytest` and its dependencies (`attrs`, `importlib-metadata`, `more-itertools`, `pluggy`, `py`, `zipp`). Each dependency is an extra burden.

> > Is there any deep reason why tests are included inside the rpy2 module and not in a separated test/ or tests/ folder? Most users do not want to install them.
> 
> Most users ? I am happy to look at data about rpy2 users if you are able to share them.
> There is of course this (#542), but otherwise I do think that it is a good idea to ship the tests because of the dependency on R. The R installation, or the R in the PATH, can change yet a user may want to check whether it is still clearing the tests.

Right, I do not have data. I am also curious how many users actually run the tests! My guess: 1%? Among them, how many are developers and have cloned the `rpy2` repo: 95%?

If the R installation changes you typically want to reinstall rpy2. I try to suggest that tests should be run (at most) at installation time.

Here I have the point of view of a package manager. Developing and distributing a software are different tasks. The tests are mostly run by package managers who check that indeed the chain of dependency of `rpy2`, including `R`, correctly fit together.

**Comment by lgautier** [comment]
I see your point but consider that "`pytest` is not a runtime dependency" until one wants to test the installation. I have had numerous issue reports that turned out to be originating from unusual and creative system setups, or API-breaking changes in R. In some cases, the installation was under the control of the user (e.g., Uni. sysadmin installing software for students). In all cases, R might be updated to a new version semi-automatically (e.g., most distros will say "hey, you have N packages ready for update. Do you want update all of them now ?") but this will not trigger an update of rpy2. Having the tests shipped along helps pinpoint the issue with minimal effort.

Now a question for you might be: shouldn't the package management process be easier ? Chasing dependencies in Python packages might cost more energy over time than lowering the burden of packaging. For example, `numpy` does ship with a `tests` module (although they have make the choice to not install the dependencies with a regular install).

**Comment by videlec** [comment]
> I see your point but consider that "`pytest` is not a runtime dependency" until one wants to test the installation. I have had numerous issue reports that turned out to be originating from unusual and creative system setups, or API-breaking changes in R. In some cases, the installation was under the control of the user (e.g., Uni. sysadmin installing software for students). In all cases, R might be updated to a new version semi-automatically (e.g., most distros will say "hey, you have N packages ready for update. Do you want update all of them now ?") but this will not trigger an update of rpy2. Having the tests shipped along helps pinpoint the issue with minimal effort.

Precisely, these "creative setups" are problematic. If `rpy2` was handled by the distro it would be updated by the package manager. Problems start when for example `R` was installed from the package manager and `rpy2` was installed with `pip`. As you said, the latter configuration is a nightmare to maintain as you have to manually chase dependencies when updating. Here having tests inside `rpy2` only helps in making diagnosis when an installation turned wrong.

For the creative system administrators my advice would just be: never install `rpy2` with pip. Use your package manager (or conda, or homebrew). The Python package manager was never designed to handle dependencies beyond pure Python packages. That is also what scipy advertize.

> Now a question for you might be: shouldn't the package management process be easier ? Chasing dependencies in Python packages might cost more energy over time than lowering the burden of packaging. For example, `numpy` does ship with a `tests` module (although they have make the choice to not install the dependencies with a regular install).

We try to make package management in SageMath as easy as possible. But since the information in `setup.py` is not reliable (eg `rpy2` depends on `R`) one needs to manually figure out what are the actual dependencies of Python packages. It is not a lot of work at start (it might have cost me less time doing it rather than discuss the matter in this issue). But as far as updates are concerned, SageMath does not have bots that would send us a notice when package updates are available (conda is great on that side).

**Comment by fcollonval** [comment]
I'm supporting @videlec here.

Shipping unit tests within the package is indeed a very good practice. But the package should not depend on test packages.

There is unfortunately no recommended way for Python for specifying the test dependencies. One possibility is to use the `extra_requires` argument in setup to define [test dependencies](https://github.com/palantir/python-language-server/blob/a91a257d2c8687a7931721d387b2ffeb6aa71fc2/setup.py#L75):

```
    extras_require={
        'test': ['versioneer',
                 'pylint>=2.5.0' if sys.version_info.major >= 3 else 'pylint',
                 'pytest', 'mock', 'pytest-cov', 'coverage', 'numpy', 'pandas',
                 'matplotlib', 'pyqt5;python_version>="3"', 'flaky'],
    }
```

So people can install the test dependencies with:

```
pip install rpy2[test]
```

---
Reason i got here is that the conda package for `rpy2` is not including `pytest` as dependency. But a good practice when building conda package is to run `pip check` to ensure pip and conda are consistent. When a conda package depends on `rpy2`, this result as an error.

**Comment by lgautier** [comment]
The advice to run `pip check` is something for conda packagers rather than for the rpy2 project. Conda/conda packaging of `rpy2`, or R, is the frequent source of issues for rpy2 users, and "use `pip install rpy2`" instead is often the answer. Here I have not idea why they do not include a required package (`pytest`).

However, I hear you all. I am seeing `pytest` as just one of the dependencies for `rpy2`, the same way that `tzlocal` is required. If never using rpy2 with datetime data, it is arguably not required. On the other hand I also see that testing can be considered somewhat different in principle. As there are more reports about requiring an issue `pytest`, and issues that are with conda or conda packaging land here, I should look into making `pytest` an optional requirement. `pandas` and `numpy` are already optional (`rpy2` can work without them).

**Comment by lgautier** [comment]
This fixed with PR#766, and in rpy2-3.4.2 (just released and on pypi).
